docs/docs/reference/implicit-function-types-spec.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: doc-page
+title: "Implicit Function Types - More Details"
+---
+
+Initial implementation in (#1775)[https://github.com/lampepfl/dotty/pull/1775].
+
+## Syntax
+
+Type              ::=  [`implicit'] FunArgTypes `=>' Type
+                    |  HkTypeParamClause `=>' Type
+                    |  InfixType
+Expr              ::=  [`implicit'] FunParams `=>' Expr
+BlockResult       ::=  [`implicit'] FunParams `=>' Block
+                    |  Expr1
+
+Implicit function types associate to the right, e.g.
+`implicit S ⇒ implicit T ⇒ U` is the same as `implicit S ⇒ (implicit T ⇒ U)`.
+
+## Implementation
+
+Implicit function types are shorthands for class types that define `apply`
+methods with implicit parameters. Specifically, the `N`-ary function type
+`implicit T1, ..., TN ⇒ R` is a shorthand for the class type
+`ImplicitFunctionN[T1 , ... , TN, R]`. Such class types are defined in the
+Scala library for `N` between 1 and 22 as follows.
+
+    package scala
+    trait ImplicitFunctionN[-T1 , ... , -TN, +R] {
+      def apply(implicit x1: T1 , ... , xN: TN): R
+    }
+
+Anonymous implicit functions `implicit (x1: T1, ..., xn: Tn) => e` map
+implicit parameters `xi` of types `Ti` to a result given by expression `e`.
+The scope of each implicit parameter `xi` is `e`. Implicit parameters must
+have pairwise distinct names.
+
+If the expected type of the anonymous implicit function is of the form
+`scala.ImplicitFunctionN[S1, ..., Sn, R]`, the expected type of `e` is `R` and
+the type `Ti` of any of the parameters `xi` can be omitted, in which case `Ti
+= Si` is assumed. If the expected type of the anonymous implicit function is
+some other type, all implicit parameter types must be explicitly given, and
+the expected type of `e` is undefined. The type of the anonymous implicit
+function is `scala.ImplicitFunctionN[S1, ...,Sn, T]`, where `T` is the packed
```

Comment:
The term "packed" does not exist anymore in Scala 3. The term to use is probably something like "widened" (i.e singleton type are widened), but we'd need to define that centrally for the new spec.